File: ui/static/js/core/specialty.js
/**
 * BossMod AI — specialty matching and the honesty rules around "done".
 *
 * The third of the three modules utils.js became. Two related jobs, one
 * vocabulary: inferring what family of work a task is and what family an
 * agent's role is (so an assignment can warn before it goes wrong), and
 * stating what a checkable done claim looks like.
 *
 * The done-claim copy is behaviourally load-bearing — `test_role_contracts.py`
 * asserts it reaches the Board's task detail and the desk's task cards, and
 * `core/tasking/board.py` produces the server half of the same rule. This is
 * the client's single copy of it.
 */
const BossModSpecialty = (() => {

    const SPECIALTY_PAIRS = [
        { work: ['write', 'draft', 'document', 'docs', 'copy', 'author', 'edit'], roles: ['writer', 'writing', 'editor', 'docs', 'author'], family: 'write' },
        { work: ['review', 'audit', 'test', 'qa', 'inspect'], roles: ['reviewer', 'auditor', 'qa', 'tester', 'review'], family: 'review' },
        { work: ['implement', 'code', 'build', 'fix', 'debug'], roles: ['engineer', 'developer', 'coder', 'eng'], family: 'implement' },
        { work: ['research', 'analyze', 'analysis'], roles: ['researcher', 'analyst'], family: 'research' },
        { work: ['design', 'mockup', 'ux'], roles: ['designer', 'design', 'ux'], family: 'design' },
    ];
    const SPECIALTY_CONFLICTS = {
        write: ['review'],
        review: ['write', 'design'],
        design: ['review', 'implement'],
        implement: ['design'],
    };
    const FINISH_LINE_DEFAULTS = {
        write: 'A named draft or document exists. Empty done does not count.',
        review: 'A checkable allow/deny (or tests/artifact) exists. Empty done does not count.',
        implement: 'Tests evidence or a named artifact exists. Empty done does not count.',
        research: 'A named findings note exists. Empty done does not count.',
        design: 'A named mockup or design file exists. Empty done does not count.',
        coordinate: 'A named plan or status note exists. Empty done does not count.',
    };
    const FALLBACK_FINISH_LINE = 'A checkable claim exists (tests, artifact, or allow/deny). Empty done does not count.';

    /**
     * The family of work a task's wording describes.
     *
     * @param {string|null} title
     * @param {string|null} description
     * @returns {string|null} null when the text names none or more than one —
     *   an ambiguous guess is worse than no guess, because everything
     *   downstream treats a family as a fact.
     */
    function inferWorkFamily(title, description) {
        const text = `${title || ''} ${description || ''}`.toLowerCase();
        if (!text.trim()) return null;
        const hits = [];
        for (const pair of SPECIALTY_PAIRS) {
            if (pair.work.some(word => text.includes(word))) hits.push(pair.family);
        }
        const unique = [...new Set(hits)];
        return unique.length === 1 ? unique[0] : null;
    }

    /**
     * The family an agent's role belongs to.
     *
     * @param {string|null} role
     * @returns {string|null} 'coordinate' for lead-shaped roles, which match
     *   everything and so never warn; null when ambiguous.
     */
    function specialtyFamily(role) {
        const text = (role || '').toLowerCase();
        if (!text) return null;
        if (['lead', 'pm', 'manager', 'coordinator', 'owner', 'director'].some(word => text.includes(word))) {
            return 'coordinate';
        }
        const hits = [];
        for (const pair of SPECIALTY_PAIRS) {
            if (pair.roles.some(word => text.includes(word))) hits.push(pair.family);
        }
        const unique = [...new Set(hits)];
        return unique.length === 1 ? unique[0] : null;
    }

    /**
     * A starting done/fail bar for a role, for the hire form to offer.
     *
     * @param {string|null} specialty
     * @param {string|null} description
     * @returns {string} Always a usable sentence; there is no empty suggestion.
     */
    function suggestFinishLine(specialty, description) {
        const family = specialtyFamily(specialty) || inferWorkFamily(null, description);
        return FINISH_LINE_DEFAULTS[family] || FALLBACK_FINISH_LINE;
    }

    /**
     * Does this agent's specialty fit this work?
     *
     * @param {string|null} role
     * @param {string|null} title
     * @param {string|null} description
     * @returns {'match'|'mismatch'|'unknown'} 'unknown' whenever either side is
     *   ambiguous, so the operator is warned only about a real conflict.
     */
    function specialtyMatch(role, title, description) {
        const work = inferWorkFamily(title, description);
        const family = specialtyFamily(role);
        if (!work || !family || family === 'coordinate') return 'unknown';
        if (family === work) return 'match';
        if ((SPECIALTY_CONFLICTS[family] || []).includes(work)) return 'mismatch';
        return 'unknown';
    }

    /**
     * A sort key: matches first, mismatches last.
     *
     * @param {object} agent
     * @param {string|null} title
     * @param {string|null} description
     * @returns {number} 0 match, 1 unknown, 2 mismatch.
     */
    function specialtyRank(agent, title, description) {
        const status = specialtyMatch(agent?.role, title, description);
        if (status === 'match') return 0;
        if (status === 'mismatch') return 2;
        return 1;
    }

    const WORK_FAMILY_LABELS = {
        write: 'writing',
        review: 'review/audit',
        implement: 'implementation',
        research: 'research',
        design: 'design',
    };

    /**
     * What to tell the operator before a mismatched assignment.
     *
     * @param {object} agent
     * @param {string|null} title
     * @param {string|null} description
     * @returns {string} '' for anything but a mismatch — the caller renders no
     *   warning rather than an empty one.
     */
    function specialtyWarningMessage(agent, title, description) {
        const status = specialtyMatch(agent?.role, title, description);
        if (status !== 'mismatch') return '';
        const name = agent?.name || 'This assignee';
        const role = agent?.role || 'unspecified specialty';
        const work = inferWorkFamily(title, description);
        const workLabel = WORK_FAMILY_LABELS[work] || 'this work';
        return `${name} is "${role}"; this work looks like ${workLabel}. Prefer a matching teammate, or assign anyway (you will be asked to confirm).`;
    }

    /**
     * What "complete" requires for one task.
     *
     * The server's own guidance wins when it sent any; otherwise this
     * reconstructs it from the work contract and the assignee's done/fail bar,
     * so the operator sees the same rule the engine will enforce.
     *
     * @param {object} task
     * @returns {string}
     */
    function doneClaimGuidance(task) {
        if (task?.done_claim_guidance) return task.done_claim_guidance;
        const bar = (task?.assigned_to_done_fail_bar || '').trim();
        const hasFiles = Boolean(task?.work_contract?.deliverables?.length);
        let base;
        if (hasFiles) {
            base = 'Complete requires the work-contract file path to exist (that file is the checkable claim).';
        } else {
            base = 'Complete/deliver requires a checkable claim: tests evidence, an artifact path that exists, or an allow/deny proof summary. Empty done is rejected.';
        }
        return bar ? `${base} What done looks like for this agent: ${bar}` : base;
    }

    /**
     * One done claim on a single line.
     *
     * @param {object|null} claim
     * @returns {string} '' when there is no claim; the type alone when it
     *   carries neither a path nor evidence.
     */
    function formatDoneClaim(claim) {
        if (!claim || typeof claim !== 'object') return '';
        const type = String(claim.type || 'proof').trim() || 'proof';
        const path = String(claim.path || '').trim();
        const evidence = String(claim.evidence || '').trim();
        if (path && evidence) return `${type} — ${path} — ${evidence}`;
        if (path) return `${type} — ${path}`;
        if (evidence) return `${type} — ${evidence}`;
        return type;
    }

    return {
        inferWorkFamily,
        specialtyFamily,
        suggestFinishLine,
        specialtyMatch,
        specialtyRank,
        specialtyWarningMessage,
        doneClaimGuidance,
        formatDoneClaim,
    };
})();

File: core/agent_loop/specialty.py
"""Pure specialty-family and finish-line inference over plain strings.

This module exists to break an import cycle, and must stay a leaf to keep it
broken. ``core.agent_pack.schema`` needs exactly one thing from the role
contracts — ``suggest_finish_line`` — but ``core.agent_loop.role_contracts``
also reaches ``core.agent_loop.deliverables`` -> ``db`` -> ``core.bm_cli`` ->
``core.tasking.board``, which imports back into ``role_contracts``. Importing
``core.agent_pack`` from a cold interpreter therefore failed unless ``db``
happened to be imported first. Pulling the string-only half down here removes
``core.agent_pack``'s edge into that cycle entirely.

So: **do not merge this back into** ``role_contracts`` **and do not import
anything from** ``db``, ``core.agent_loop``, ``core.tasking`` **or**
``core.bm_cli`` **here.** Everything in this module is a pure function of the
strings passed to it — no ``Agent``, no ``Task``, no database, no filesystem.
Anything needing those belongs in ``role_contracts`` instead.

``role_contracts`` re-exports every public name below, so existing importers
(and ``tests/test_role_contracts.py``) keep working unchanged.
"""

from __future__ import annotations

import re
from typing import Literal

SpecialtyFamily = Literal["write", "review", "implement", "research", "design", "coordinate"]

_TOKEN_RE = re.compile(r"[a-z0-9]+")

_SPECIALTY_TOKENS: dict[SpecialtyFamily, frozenset[str]] = {
    "write": frozenset({
        "writer", "writing", "copy", "copywriter", "docs", "documentation",
        "documenter", "author", "editor", "draft",
    }),
    "review": frozenset({
        "reviewer", "review", "auditor", "audit", "qa", "tester", "test",
        "testing", "inspector", "inspect",
    }),
    "implement": frozenset({
        "engineer", "eng", "developer", "dev", "coder", "programmer",
        "implement", "implementation", "code", "coding", "builder",
    }),
    "research": frozenset({
        "researcher", "research", "analyst", "analysis", "investigate",
    }),
    "design": frozenset({
        "designer", "design", "ux", "ui", "mockup",
    }),
    "coordinate": frozenset({
        "pm", "product", "manager", "lead", "coordinator", "owner", "director",
    }),
}

_WORK_KIND_TOKENS: dict[SpecialtyFamily, frozenset[str]] = {
    "write": frozenset({"write", "writing", "draft", "document", "docs", "copy", "author", "edit", "edits"}),
    "review": frozenset({"review", "audit", "auditing", "qa", "test", "tests", "testing", "inspect"}),
    "implement": frozenset({"implement", "code", "coding", "build", "fix", "debug", "develop"}),
    "research": frozenset({"research", "analyze", "analysis", "investigate"}),
    "design": frozenset({"design", "mockup", "wireframe", "ux"}),
}

# Operator-facing finish-line defaults. Empty done stays blocked by the
# checkable-claim rules even when the stored bar is blank.
_DEFAULT_FINISH_LINES: dict[SpecialtyFamily, str] = {
    "write": "A named draft or document exists. Empty done does not count.",
    "review": "A checkable allow/deny (or tests/artifact) exists. Empty done does not count.",
    "implement": "Tests evidence or a named artifact exists. Empty done does not count.",
    "research": "A named findings note exists. Empty done does not count.",
    "design": "A named mockup or design file exists. Empty done does not count.",
    "coordinate": "A named plan or status note exists. Empty done does not count.",
}
_FALLBACK_FINISH_LINE = (
    "A checkable claim exists (tests, artifact, or allow/deny). Empty done does not count."
)


def tokenize(text: str | None) -> set[str]:
    """Lowercase alphanumeric tokens from free text.

    Args:
        text: Any free-form string, or ``None``.

    Returns:
        The set of lowercased ``[a-z0-9]+`` runs in ``text``; an empty set for
        ``None`` or an empty/punctuation-only string.
    """
    if not text:
        return set()
    return set(_TOKEN_RE.findall(text.lower()))


def specialty_family(role: str | None) -> SpecialtyFamily | None:
    """Map a hire specialty (``Agent.role``) onto one v1 family, if clear.

    Args:
        role: The one-line hire specialty text, or ``None``.

    Returns:
        The single matching family, or ``None`` when the text is empty, matches
        no family, or matches more than one specific family. Ambiguity is
        reported as ``None`` rather than resolved by guessing, because callers
        treat an unknown family as "no opinion" instead of acting on it.
        ``"coordinate"`` only wins when nothing more specific matched.
    """
    tokens = tokenize(role)
    if not tokens:
        return None
    hits = [
        family
        for family, family_tokens in _SPECIALTY_TOKENS.items()
        if tokens & family_tokens
    ]
    specific = [family for family in hits if family != "coordinate"]
    if len(specific) == 1:
        return specific[0]
    if len(specific) > 1:
        return None
    if hits == ["coordinate"]:
        return "coordinate"
    return None


def infer_work_kind(
    title: str | None,
    description: str | None = None,
    *,
    requested_specialty: str | None = None,
) -> SpecialtyFamily | None:
    """Infer work kind from an explicit requested specialty or title/description.

    An explicit ``requested_specialty`` wins outright unless it resolves to
    ``"coordinate"``, which is too broad to describe a unit of work.

    Args:
        title: Task title, or ``None``.
        description: Task description, or ``None``.
        requested_specialty: Specialty the caller asked for, if any.

    Returns:
        The inferred family, or ``None`` when there is no signal or when two
        families tie on token overlap.
    """
    requested = specialty_family(requested_specialty)
    if requested and requested != "coordinate":
        return requested
    tokens = tokenize(title) | tokenize(description)
    if not tokens:
        return requested
    scores: dict[SpecialtyFamily, int] = {}
    for family, family_tokens in _WORK_KIND_TOKENS.items():
        overlap = tokens & family_tokens
        if overlap:
            scores[family] = len(overlap)
    if not scores:
        return requested
    best = max(scores.values())
    winners = [family for family, score in scores.items() if score == best]
    if len(winners) == 1:
        return winners[0]
    return None


def suggest_finish_line(
    specialty: str | None,
    description: str | None = None,
) -> str:
    """Suggest a default finish line from specialty, using description when useful.

    Does not persist or rewrite stored text. Blank ``done_fail_bar`` stays valid;
    empty done is still rejected by the checkable-claim rules.

    Args:
        specialty: The hire specialty text, or ``None``.
        description: Free-text description used only when ``specialty`` does not
            resolve to a family on its own.

    Returns:
        Always a non-empty operator-facing line: the family default, or the
        generic checkable-claim fallback when no family could be inferred.
    """
    family = specialty_family(specialty)
    if family is None:
        family = infer_work_kind(None, description)
    if family is None:
        return _FALLBACK_FINISH_LINE
    return _DEFAULT_FINISH_LINES[family]
